File: 43.字符串相乘.go
/*
 * @lc app=leetcode.cn id=43 lang=golang
 *
 * [43] 字符串相乘
 */
package main

import "strconv"

// @lc code=start

func multiply(num1 string, num2 string) string {
	if num1 == "0" || num2 == "0" {
		return "0"
	}
	res := "0"
	len1, len2 := len(num1), len(num2)
	for i := len2 - 1; i >= 0; i-- {
		curr := ""
		add := 0
		for j := len2 - 1; j > i; j-- {
			curr += "0"
		}
		y := int(num2[i] - '0')
		for j := len1 - 1; j >= 0; j-- {
			x := int(num1[j] - '0')
			product := x*y + add
			curr = strconv.Itoa(product%10) + curr
			add = product / 10
		}
		for ; add != 0; add /= 10 {
			curr = strconv.Itoa(add%10) + curr
		}
		res = addStrings(res, curr)
	}
	return res
}

func addStrings(num1 string, num2 string) string {
	index1, index2 := len(num1)-1, len(num2)-1
	add := 0
	res := ""
	for ; index1 >= 0 || index2 >= 0 || add != 0; index1, index2 = index1-1, index2-1 {
		x, y := 0, 0
		if index1 >= 0 {
			x = int(num1[index1] - '0')
		}
		if index2 >= 0 {
			y = int(num2[index2] - '0')
		}
		tempRes := x + y + add
		res = strconv.Itoa(tempRes%10) + res
		add = tempRes / 10
	}
	return res
}

// @lc code=end

File: 43.字符串相乘.java
/*
 * @lc app=leetcode.cn id=43 lang=java
 *
 * [43] 字符串相乘
 */

// @lc code=start
class Solution {
    public String multiply(String num1, String num2) {
        if (num1.equals("0") || num2.equals("0")) {
            return "0";
        }
        StringBuilder res = new StringBuilder();
        int len1 = num1.length();
        int len2 = num2.length();
        int[] resArr = new int[len1 + len2];

        for (int i = len1 - 1; i >= 0; i--) {
            int x = num1.charAt(i) - '0';
            for (int j = len2 - 1; j >= 0; j--) {
                int y = num2.charAt(j) - '0';
                resArr[i + j + 1] += x * y;
            }
        }

        for (int i = len1 + len2 - 1; i > 0; i--) {
            resArr[i - 1] += resArr[i] / 10;
            resArr[i] %= 10;
        }

        int begin = resArr[0] == 0 ? 1 : 0;
        for (int i = begin; i < len1 + len2; i++) {
            res.append(resArr[i]);
        }
        return res.toString();
    }
}
// @lc code=end

